File: samples/ObjC Manual Dictionary Sample/Sample/GeofenceVC.m
/*
 * Licensed Materials - Property of IBM
 *
 * 5725E28, 5725I03
 *
 * © Copyright IBM Corp. 2016, 2019
 * US Government Users Restricted Rights - Use, duplication or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 */

#import "GeofenceVC.h"
#import <IBMMobilePush/IBMMobilePush.h>

@interface GeofenceVC ()
@property CLLocationManager * locationManager;
@property CLLocation * lastLocation;
@property BOOL followGPS;
@property NSMutableSet * overlayIds;
@property dispatch_queue_t queue;
@property NSMutableDictionary * circleToIdentifier;
@end

@implementation GeofenceVC
-(void)awakeFromNib
{
    [super awakeFromNib];
    self.queue = dispatch_queue_create("background", nil);
    self.status.accessibilityIdentifier=@"status";
}

-(IBAction)enable:(id)sender
{
    [MCESdk.sharedInstance manualLocationInitialization];
}

-(void)updateStatus
{
    MCEConfig* config = [[MCESdk sharedInstance] config];
    if(config.geofenceEnabled)
    {
        switch(CLLocationManager.authorizationStatus)
        {
            case kCLAuthorizationStatusDenied:
            [self.status setTitle:@"DENIED" forState:UIControlStateNormal];
            [self.status setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
            break;
            case kCLAuthorizationStatusNotDetermined:
            [self.status setTitle:@"DELAYED (Touch to enable)" forState:UIControlStateNormal];
            [self.status setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
            break;
            case kCLAuthorizationStatusAuthorizedAlways:
            [self.status setTitle:@"ENABLED" forState:UIControlStateNormal];
            [self.status setTitleColor:[UIColor greenColor] forState:UIControlStateNormal];
            break;
            case kCLAuthorizationStatusRestricted:
            [self.status setTitle:@"RESTRICTED?" forState:UIControlStateNormal];
            [self.status setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
            
            break;
            case kCLAuthorizationStatusAuthorizedWhenInUse:
            [self.status setTitle:@"ENABLED WHEN IN USE" forState:UIControlStateNormal];
            [self.status setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
            break;
        }
    }
    else
    {
        [self.status setTitle:@"DISABLED" forState:UIControlStateNormal];
        [self.status setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
    }
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self createMonitor];
    [self updateStatus];
}

-(void)createMonitor
{
    self.overlayIds = [NSMutableSet set];
    self.circleToIdentifier = [NSMutableDictionary dictionary];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.locationManager = [[CLLocationManager alloc]init];
        self.locationManager.delegate = self;
        [self.locationManager startUpdatingLocation];
    });
}

-(void)viewDidDisappear:(BOOL)animated
{
    [self destroyMonitor];
}
-(void)destroyMonitor
{
    [self.mapView removeOverlays: self.mapView.overlays];
    self.overlayIds = nil;
    [self.locationManager stopUpdatingLocation];
    self.locationManager.delegate = nil;
    self.locationManager = nil;
}

-(IBAction)refresh:(id)sender
{
    [[[MCELocationClient alloc] init] scheduleSync];
}

-(void) viewDidLoad
{
    [super viewDidLoad];
    self.followGPS=TRUE;
    [self updateGpsButton];
    UIPanGestureRecognizer* panRec = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(didDragMap:)];
    [panRec setDelegate:self];
    [self.mapView addGestureRecognizer:panRec];
    
    self.mapView.showsUserLocation = TRUE;
    self.mapView.delegate=self;
    
    [[NSNotificationCenter defaultCenter] addObserverForName:@"RefreshActiveGeofences" object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification * _Nonnull note) {
        [self addGeofenceOverlaysNearCoordinate:self.lastLocation.coordinate radius:100000];
        NSArray * overlays = self.mapView.overlays;
        [self.mapView removeOverlays:overlays];
        [self.mapView addOverlays:overlays];
    }];
    
    [[NSNotificationCenter defaultCenter] addObserverForName:@"DownloadedLocations" object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification * _Nonnull note) {
        
        [self addGeofenceOverlaysNearCoordinate: self.lastLocation.coordinate radius:100000];
    }];
}

- (void)didDragMap:(UIGestureRecognizer*)gestureRecognizer
{
    if (gestureRecognizer.state == UIGestureRecognizerStateEnded)
    {
        self.followGPS=FALSE;
        [self updateGpsButton];
        
        MKCoordinateRegion region = self.mapView.region;
        CLLocation * location = [[CLLocation alloc]initWithLatitude: region.center.latitude longitude: region.center.longitude];
        
        [self addGeofenceOverlaysNearCoordinate: location.coordinate radius:100000];
        self.lastLocation = location;
    }
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer
{
    return YES;
}

-(IBAction)clickGpsButton:(id)sender
{
    self.followGPS = !self.followGPS;
    [self updateGpsButton];
}

-(void)updateGpsButton
{
    if(self.followGPS)
    self.gpsButton.tintColor=[UIColor colorWithRed:0 green:0.4784313725 blue:1 alpha:1];
    else
    self.gpsButton.tintColor=[UIColor colorWithRed:0 green:0.4784313725 blue:1 alpha:0.3];
}

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray<CLLocation *> *)locations
{
    CLLocation * location = locations.lastObject;
    if(self.followGPS)
    {
        MKCoordinateRegion region = MKCoordinateRegionMakeWithDistance(location.coordinate, 1000, 1000);
        [self.mapView setRegion:region animated:TRUE];

        if(self.lastLocation==nil || [self.lastLocation distanceFromLocation:location] > 10)
        {
            [self addGeofenceOverlaysNearCoordinate:location.coordinate radius:10000];
            self.lastLocation = location;
        }
    }
}

-(void)addGeofenceOverlaysNearCoordinate: (CLLocationCoordinate2D) coordinate radius: (double) radius
{
    dispatch_async(dispatch_get_main_queue(), ^(void){
        NSMutableArray * additionalOverlays = [NSMutableArray array];
        NSMutableArray * removeOverlays = [NSMutableArray array];
        NSMutableSet * geofences = [[MCELocationDatabase sharedInstance] geofencesNearCoordinate:coordinate radius:radius];
        
        NSMutableSet * currentlyDisplayedIds = [NSMutableSet set];
        for(MKCircle * overlay in self.mapView.overlays) {
            [currentlyDisplayedIds addObject:overlay.title];
        }
        
        NSMutableSet * currentGeofenceIds = [NSMutableSet set];
        for (MCEGeofence * geofence in geofences) {
            [currentGeofenceIds addObject: geofence.locationId];
            if( ![currentlyDisplayedIds containsObject: geofence.locationId] ) {
                MKCircle * circle = [MKCircle circleWithCenterCoordinate:geofence.coordinate radius:geofence.radius];
                circle.title = geofence.locationId;
                [additionalOverlays addObject: circle];
            }
        }
        
        for(MKCircle * overlay in self.mapView.overlays) {
            if(![currentGeofenceIds containsObject:overlay.title]) {
                [removeOverlays addObject:overlay];
            }
        }
        
        [self.mapView removeOverlays: removeOverlays];
        [self.mapView addOverlays: additionalOverlays];
    });
}

- (MKOverlayRenderer *)mapView:(MKMapView *)mapView rendererForOverlay:(id<MKOverlay>)overlay
{
    BOOL active = false;
    if([overlay isKindOfClass:[MKCircle class]])
    {
        MKCircle * circle = (MKCircle *)overlay;
        
        NSString * identifier = circle.title;
        
        if(identifier)
        {
            for (CLRegion * region in self.locationManager.monitoredRegions) {
                if([region.identifier isEqual: identifier])
                {
                    active=true;
                }
            }
        }
    }
    MKCircleRenderer * renderer = [[MKCircleRenderer alloc] initWithCircle: overlay];
    
    if(active)
    {
        renderer.fillColor = [UIColor colorWithRed:1 green:0 blue:0 alpha:0.1];
        renderer.strokeColor = [UIColor colorWithRed:1 green:0 blue:0 alpha:1.0];
    }
    else
    {
        renderer.fillColor = [UIColor colorWithRed:0 green:0.4784313725 blue:1 alpha:0.1];
        renderer.strokeColor = [UIColor colorWithRed:0 green:0.4784313725 blue:1 alpha:1.0];
    }
    renderer.lineWidth = 1;
    renderer.lineDashPattern = @[ @(2), @(2) ];
    return renderer;
}

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
    [self updateStatus];
}

@end

File: samples/SwiftSample/GeofenceVC.swift
/*
 * Licensed Materials - Property of IBM
 *
 * 5725E28, 5725I03
 *
 * © Copyright IBM Corp. 2016, 2016
 * US Government Users Restricted Rights - Use, duplication or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 */

import UIKit
import MapKit

class GeofenceVC: UIViewController, CLLocationManagerDelegate, MKMapViewDelegate, UIGestureRecognizerDelegate {
    
    @IBOutlet weak var mapView: MKMapView!
    @IBOutlet weak var gpsButton: UIBarButtonItem!
    @IBOutlet weak var status: UILabel!
    
    var locationManager: CLLocationManager
    var lastLocation: CLLocation?
    var followGPS: Bool = true
    var overlayIds: NSMutableSet = []
    var queue: DispatchQueue
    
    required init?(coder aDecoder: NSCoder) {
        locationManager = CLLocationManager()
        queue = DispatchQueue(label: "background")
        super.init(coder: aDecoder)
        locationManager.delegate = self
        
        NotificationCenter.default.addObserver(forName: NSNotification.Name("DownloadedLocations"), object: nil, queue: OperationQueue.main) { (note) in
            self.addGeofenceOverlaysNearCoordinate(coordinate: self.lastLocation!.coordinate, radius: 1000)
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        let config = MCESdk.sharedInstance().config;
        if(config!.geofenceEnabled)
        {
            status.text="ENABLED"
            status.textColor = UIColor.green
        }
        else
        {
            status.text="DISABLED"
            status.textColor = UIColor.red
        }
        
        createMonitor()
    }
    
    func createMonitor()
    {
        overlayIds = []
        let authStatus = CLLocationManager.authorizationStatus()
        if(authStatus == CLAuthorizationStatus.restricted || authStatus == CLAuthorizationStatus.denied)
        {
            NSLog("Not allowed to use location services, notify user?")
            return
        }
        
        locationManager.startUpdatingLocation()
        
        if(authStatus == CLAuthorizationStatus.notDetermined)
        {
            NSLog("Auth status unknown, asking user")
            locationManager.requestWhenInUseAuthorization()
        }
        
        if(!CLLocationManager.locationServicesEnabled())
        {
            NSLog("Location services is not enabled, notifiy user?")
            return
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        destroyMonitor()
    }
    
    func destroyMonitor()
    {
        mapView.removeOverlays(mapView.overlays)
        mapView.removeAnnotations(mapView.annotations)
        overlayIds = []
        locationManager.stopUpdatingLocation()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        followGPS=true
        updateGpsButton()
        let panRec = UIPanGestureRecognizer(target: self, action: #selector(didDragMap(gestureRecognizer:)))
        panRec.delegate=self
        mapView.addGestureRecognizer(panRec)
        mapView.showsUserLocation=true
        mapView.delegate=self
    }
    
    func didDragMap(gestureRecognizer: UIGestureRecognizer)
    {
        if(gestureRecognizer.state == UIGestureRecognizerState.ended)
        {
            followGPS=false
            updateGpsButton()
            let region = mapView.region
            let location = CLLocation(latitude: region.center.latitude, longitude: region.center.longitude)
            if lastLocation != nil || lastLocation!.distance(from: location) > 10
            {
                let north = CLLocation(latitude: region.center.latitude - region.span.latitudeDelta * 0.5, longitude: region.center.longitude)
                let south = CLLocation(latitude: region.center.latitude + region.span.latitudeDelta * 0.5, longitude: region.center.longitude)
                let metersLatitude = north.distance(from: south)
                
                let east = CLLocation(latitude: region.center.latitude, longitude: region.center.longitude - region.span.longitudeDelta * 0.5)
                let west = CLLocation(latitude: region.center.latitude, longitude: region.center.longitude + region.span.longitudeDelta * 0.5)
                let metersLongitude = east.distance(from: west)
                
                let maxMeters = max(metersLatitude, metersLongitude)
                addGeofenceOverlaysNearCoordinate(coordinate: location.coordinate, radius: maxMeters)
                lastLocation = location
            }
        }
    }
    
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool
    {
        return true
    }
    
    @IBAction func refresh(sender: AnyObject)
    {
        MCELocationClient().scheduleSync()
    }
    
    @IBAction func clickGpsButton(sender: AnyObject)
    {
        followGPS = !followGPS
        updateGpsButton()
    }
    
    func updateGpsButton()
    {
        if(followGPS)
        {
            gpsButton.tintColor = UIColor(red: 0, green: 0.4784313725, blue: 1, alpha: 1)
        }
        else
        {
            gpsButton.tintColor = UIColor(red: 0, green: 0.4784313725, blue: 1, alpha: 0.3)
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        let location = locations.last
        if(followGPS)
        {
            let region = MKCoordinateRegionMakeWithDistance(location!.coordinate, 1000, 1000)
            mapView.setRegion(region, animated: true)
        }
        
        if(lastLocation == nil || lastLocation!.distance(from: location!)>10)
        {
            addGeofenceOverlaysNearCoordinate(coordinate: location!.coordinate, radius: 1000)
            lastLocation = location
        }
    }
    
    func addGeofenceOverlaysNearCoordinate(coordinate: CLLocationCoordinate2D, radius: Double)
    {
        let r = min(radius, 10000)
        DispatchQueue.global().async {
            var overlays = [MKOverlay]()
            var annotations = [MKAnnotation]()
            let geofences = MCELocationDatabase.sharedInstance().geofencesNearCoordinate(coordinate, radius: r)
            for geofence in geofences!
            {
                if let geofence = geofence as? CLCircularRegion
                {
                    if !self.overlayIds.contains(geofence.identifier)
                    {
                        overlays.append(MKCircle(center: geofence.center, radius: geofence.radius))
                        self.overlayIds.add(geofence.identifier)
                        let annotation = MKPointAnnotation()
                        annotation.coordinate=geofence.center
                        annotation.title=geofence.identifier
                        annotation.subtitle=NSString(format: "Latitude %f, Longitude %f, Radius: %.1f", geofence.center.latitude, geofence.center.longitude, geofence.radius) as String
                        annotations.append(annotation)
                    }
                }
            }
            DispatchQueue.main.async {
                self.mapView.addOverlays(overlays)
                self.mapView.addAnnotations(annotations)
            }
        }
    }
    
    func mapView(_ mapView: MKMapView, rendererFor overlay: MKOverlay) -> MKOverlayRenderer {
        let renderer = MKCircleRenderer(overlay: overlay)
        renderer.fillColor = UIColor(red: 0, green: 0.4784313725, blue: 1, alpha: 0.1)
        renderer.strokeColor = UIColor(red: 0, green: 0.4784313725, blue: 1, alpha: 1.0)
        renderer.lineWidth = 1
        renderer.lineDashPattern = [2,2]
        return renderer
    }
}
